stable_baselines3/her/her.py

```diff
@@ -0,0 +1,495 @@
+import io
+import pathlib
+from typing import Callable, Iterable, List, Optional, Tuple, Type, Union
+
+import gym
+import numpy as np
+
+from stable_baselines3.common.base_class import BaseAlgorithm
+from stable_baselines3.common.buffers import ReplayBuffer
+from stable_baselines3.common.callbacks import BaseCallback
+from stable_baselines3.common.noise import ActionNoise
+from stable_baselines3.common.off_policy_algorithm import OffPolicyAlgorithm
+from stable_baselines3.common.policies import BasePolicy
+from stable_baselines3.common.save_util import load_from_zip_file, recursive_getattr, recursive_setattr
+from stable_baselines3.common.type_aliases import GymEnv, MaybeCallback, RolloutReturn
+from stable_baselines3.common.utils import check_for_correct_spaces
+from stable_baselines3.common.vec_env import VecEnv
+from stable_baselines3.common.vec_env.obs_dict_wrapper import ObsDictWrapper
+from stable_baselines3.her.goal_selection_strategy import KEY_TO_GOAL_STRATEGY, GoalSelectionStrategy
+from stable_baselines3.her.her_replay_buffer import HerReplayBuffer
+
+
+def get_time_limit(env: VecEnv, current_max_episode_length: Optional[int]) -> int:
+    """
+    Get time limit from environment.
+
+    :param env: Environment from which we want to get the time limit.
+    :param current_max_episode_length: Current value for max_episode_length.
+    :return: max episode length
+    """
+    # try to get the attribute from environment
+    if current_max_episode_length is None:
+        try:
+            current_max_episode_length = env.get_attr("spec")[0].max_episode_steps
+        # if not available check if a valid value was passed as an argument
+        except AttributeError:
+            raise ValueError(
+                "The max episode length could not be inferred."
+                "You must specify a `max_episode_steps` when registering the environment, "
+                "use a `gym.wrappers.TimeLimit` wrapper "
+                "or pass `max_episode_length` to the model constructor"
+            )
+    return current_max_episode_length
+
+
+class HER(BaseAlgorithm):
+    """
+    Hindsight Experience Replay (HER)
```

Comment:
maybe add a warning in the doc saying that we require the env or the user to provide a maximum episode length
and also say that to refer to the offline algos doc for the additional arguments